### ISSUE TITLE
Proper implementation of RecoveryAccessor read operations

### ISSUE THREAD

**Issue by killenb** [ChimeraTK/ApplicationCore#138]
Child of #123 

I had to split it into two phases. Phase 1 is needed for #130, Phase 2 depends on #130 and #131

Note: Phase 1 has already been implemented!

Description of phase 2:
The implementation of #130 adapted to the changes in DeviceAccess (exceptions are not thrown in the transfers but in the post actions), but it does not fully implement the new specification yet. There are three missing features:
* The first read() or readAsync() on a bad device goes through with old data and DataValidity::invalid. Only the second read()/readAsync() blocks
* readNonBlocking() /readLatest() never block
* the initial value is never transported

In addition there is the requirement that no new transfers must be started when the device is recovering, which means the non-blocking reads, which obviously cannot call DeviceModule::waitForRecovery() because it blocks, must not simply try to perform the transfer and let it fail.

Implementation
--------------------

If nothing else is mentioned it is phase 1 and stays in phase 2.

* Create a protected variable `transferAllowed`
* [ ] doPreRead()
  * Don't call wait for recovery here
  * **Phase 1 only** Set transferAllowed to true if the application is in LifeCycleState::run. Otherwise set it to false.
  * **Phase 2** Call DeviceModule::startTransfer(). The return value tells you if the device is functional and you can execute the transfer. Store it in `tranferAllowed`.
* [ ] doReadTransferXxx()
  ** Only call _impl->readTransferXxx() if `transferAllowed` is true. If not return with `hasNewData` as false.
* [ ] doPostRead()
  *  Don't use the genericPostAction() any more.
  * **Phase2** If (transferAllowed) DeviceModule::stopTransfer();
  *  Try {_impl->postRead()} and catch the exception. If there was an exception
    * report it to the device module
    * set a local variable `hasException` to true
  * A decision depends on the phase
    * **Phase 1** If  (`hasException` or `!transferAllowed`)
    * **Pase 2** If ((`hasException` or `!transferAllowed`) and `previousReadFailed` and (TransferType==read or TranferType==readAsync)) or (versionNumber == {nullptr})
  * If decision is true repeat the the following sequence until the read succeeds (does not thow)
      * waitForRecovery()
      * **Phase 2** if (not DeviceModule::startTransfer()) go to beginning of loop
      * impl->read()
      * **Phase 2** DeviceModule::stopTransfer()
  Also set "hasException" to false once the read succeeds.
  * setOwnerValidity(hasException)

Although in phase 1 everything could be done in the catch-block and setOwnerValitiy is always called with true, please implement it like this in preparation of phase 2.

When re-trying to read we always have to use a complete read cycle incl. preRead and postRead because we cannot repeat only the transfer. We are already behind the delegated postRead and have to call the full sequence. 

Notice: Instead of `transferAllowed`you cannot use `TransferElement::readTransactionInProgress`. If this is false the doPostRead is never called.

Sorry, I don't know if it's a good sign that the description of the implementation basically is the code. I always think good code should read like the newspaper.

Definition of done:
* [ ] All tests are succeeding, especially the new ones of #155 and #156

**Comment by killenb** [comment]
> Sorry, I don't know if it's a good sign that the description of the implementation basically is the code. I always think good code should read like the newspaper.

If obviously is a bad sign. I am micro-managing the tickets too much.

**Comment by mhier** [comment]
If !transferAllowed it must IMO not wait for the recovery in postRead.If !transferAllowed we are in prepare(), hence devices will never recover at that point. That would prevent applications from starting... Am I missing something?

**Comment by killenb** [comment]
This ticket has become obsolete with the changes in the spec. The procedure described has changed. It is described in the spec, and implemented already
